pkg/controller/unreachable/unreachable_controller.go

```diff
@@ -145,88 +144,84 @@ func (r *ReconcileRemoteMachineSet) Reconcile(request reconcile.Request) (reconc
 		return reconcile.Result{}, nil
 	}
 
-	// Check if we're due for rechecking cluster's connectivity
-	cond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.UnreachableCondition)
-	if cond != nil {
-		if !isTimeForConnectivityCheck(cond) {
-			cdLog.WithFields(log.Fields{
-				"sinceLastTransition": time.Since(cond.LastTransitionTime.Time),
-				"sinceLastProbe":      time.Since(cond.LastProbeTime.Time),
-			}).Debug("skipping unreachable check")
-			return reconcile.Result{}, nil
-		}
+	wasUnreachable, lastCheck := remoteclient.Unreachable(cd)
+	wasPrimaryActive := remoteclient.IsPrimaryURLActive(cd)
+	connectivityCheckDelay := maxUnreachableDuration - time.Since(lastCheck)
+
+	if !wasUnreachable && wasPrimaryActive && connectivityCheckDelay > 0*time.Second {
+		cdLog.WithField("delay", connectivityCheckDelay).Debug("waiting to check connectivity")
+		return reconcile.Result{RequeueAfter: connectivityCheckDelay}, nil
 	}
 
 	cdLog.Info("checking if cluster is reachable")
 	remoteClientBuilder := r.remoteClusterAPIClientBuilder(cd)
-	hasOverride := cd.Spec.ControlPlaneConfig.APIURLOverride != ""
-	var (
-		unreachableError          error
-		activeAPIURLOverrideError error
-	)
-	if _, primaryErr := remoteClientBuilder.UsePrimaryAPIURL().Build(); primaryErr != nil {
-		if hasOverride {
+	var unreachableError error
+	updateUnreachable := true
+	_, primaryErr := remoteClientBuilder.UsePrimaryAPIURL().Build()
+	if primaryErr != nil {
+		if hasOverride(cd) {
 			cdLog.WithError(primaryErr).Info("unable to create remote API client using API URL override")
-			activeAPIURLOverrideError = primaryErr
-			if _, secondaryErr := remoteClientBuilder.UseSecondaryAPIURL().Build(); secondaryErr != nil {
-				cdLog.WithError(secondaryErr).Warn("unable to create remote API client with either the initial API URL or the API URL override, marking cluster unreachable")
-				unreachableError = utilerrors.NewAggregate([]error{primaryErr, secondaryErr})
+			if wasUnreachable || wasPrimaryActive || connectivityCheckDelay <= 0*time.Second {
+				if _, secondaryErr := remoteClientBuilder.UseSecondaryAPIURL().Build(); secondaryErr != nil {
```

Comment:
Got confused again at this point, throughout the PR I've seen initial, override, primary, secondary. Can these be collapsed into just two consistent terms?

Looking further maybe not. Initial and override I understand. It looks like primary is whatever is supposed to be active, and changes once the override is available. Secondary must be whatever is non-primary at that point in time?

Reply:
I think you've got it.
**Initial**: what comes from the installer in the kubeconfig
**Active**: the API URL that is currently being used to connect to the remote cluster
**Primary**: the preferred API URL. If there is an override, this is the override URL. If there is not an override, this is the initial URL.
**Secondary**: the fallback API URL. If there is an override, this is the initial URL.

Reply:
There are comments describing what Primary and Secondary mean in the `remoteclient.Builder` interface.
https://github.com/openshift/hive/blob/e503130882dce2c6945762439283e9a62d0467f8/pkg/remoteclient/remoteclient.go#L47-L53